Firmware Progress

micade145 opened the pull request:
Quick update on Amperes firmware: 
- Amperes_Init and Amperes_Task run
- Need to test with hardware to make verify ADC readings

To Do:
- Test CAN

Questions:
- how should I send the ADC reading over CAN? is there a format to put the uint32_t reading into the uint8_t array?

micade145 committed:
added Embedded sharepoint

micade145 committed:
gitignore, added placeholder files to check compilation (it compiles hooray)

micade145 committed:
started outline for driver

micade145 committed:
first iteration of amperes code, testing init and get_reading using main

micade145 committed:
ADC init seems to work, CAN is set up (not tested)

Lakshay983 commented on Firmware/Apps/Src/main.c:
```diff
@@ -0,0 +1,27 @@
+                    (void*) 1,
+                    tskIDLE_PRIORITY+4,
+                    amperes_task_stack,
+                    &amperes_task_buffer);
```
Create an init task, then put the Amperes task in there

several RTOS things only work inside the FreeRTOS scheduler, so you make an init Task, start the scheduler, spawn all ur tasks, then delete the init task.

See how BPS-Leader repo does it

I'd also recommend init'ing CAN in your init task. See CAN.h on instructions on how to properly init CAN

Lakshay983 commented on Firmware/Drivers/Inc/Amperes.h:
```diff
@@ -0,0 +1,34 @@
+#define AMPERES_CHANNEL ADC_CHANNEL_1
+#define AMPERES_SAMPLE_TIME ADC_SAMPLETIME_3CYCLES
+#endif
+
```
You don't need to support multiple types of microcontrollers imo, just use the STM32L4xx

micade145 replied:
true true I just have this for debugging (I only have an stm32F4 rn)

Lakshay983 commented on Firmware/Drivers/Inc/Amperes.h:
```diff
@@ -0,0 +1,34 @@
+#endif
+
+extern StaticTask_t amperes_task_buffer;
+extern StackType_t amperes_task_stack[configMINIMAL_STACK_SIZE];
```
if you ever want to add more RTOS tasks it's good to have all your buffers and stacks inside one file so maybe make like a Tasks.h file

If you want to do smth funny, i'm a big hater of needing to change multiple things to add smth, so for example if I wnat to create a new RTOS task i need to create a new StaticTask_t  and StackType_t . It'd be nice if it were all one line and autogenerated.
Ie I have a bunch of macros that autogenerate things I need. It's how the CAN driver generates it's table of IDs to look at.

See how I do it here in the FOREACH_DISPLAY_COMPONENT
https://github.com/lhr-solar/Controls/blob/pedals_integration/Drivers/Inc/Display.h
or here: https://github.com/lhr-solar/Controls/blob/pedals_integration/Drivers/Inc/DebugIO.h

I like it but also may not be worth so feel free to ignore. Makings things really easy to expand and change as time goes in is a good philosohpy to keep in mind as you continue to write code other people will work on and add onto. You may know exactly what you need to change but people at a first glance may not

Lakshay983 commented on Firmware/Apps/Src/main.c:
```diff
@@ -0,0 +1,27 @@
+                    "Amperes Task",
+                    configMINIMAL_STACK_SIZE,
+                    (void*) 1,
+                    tskIDLE_PRIORITY+4,
```
Make your priority a macro, also why +4?

Rav4s commented on Firmware/Apps/Src/main.c:
```diff
@@ -0,0 +1,27 @@
+#include "stm32xx_hal.h"
+#include "Amperes.h"
+
+uint32_t reading;
```
don't see this used anywhere
```suggestion
```

Rav4s commented on Firmware/Apps/Src/main.c:
```diff
@@ -0,0 +1,27 @@
+    SystemClock_Config();
+
+    // Initializing
+    if (Amperes_Init() != INIT_OK) return 0;
```
call `error_handler` here so we know an error occurred
```suggestion
    if (Amperes_Init() != INIT_OK) error_handler();
```

Rav4s commented on Firmware/Apps/Src/main.c:
```diff
@@ -0,0 +1,27 @@
+                    "Amperes Task",
+                    configMINIMAL_STACK_SIZE,
+                    (void*) 1,
+                    tskIDLE_PRIORITY+4,
```
reasoning for such a low priority? also should make this a macro

micade145 replied:
no reason I just grabbed this from the Emb sharepoint adc test

Rav4s commented on Firmware/Drivers/Src/Amperes.c:
```diff
@@ -0,0 +1,141 @@
+        // Debug
+        HAL_GPIO_TogglePin(GPIOA, GPIO_PIN_5);
+
+        vTaskDelay(200);
```
this is ticks right? should prolly make it a macro and document why this rate

Lakshay983 replied:
use pdMsToTicks

Rav4s replied:
can't use under 1ms gang

Rav4s replied:
although depending on our tick rate it prob won't matter

Rav4s commented on Firmware/Drivers/Inc/Amperes.h:
```diff
@@ -0,0 +1,34 @@
+    READING_FAIL
+} AmperesStatus_t;
+
+AmperesStatus_t Amperes_Init();
+AmperesStatus_t Amperes_GetReading(uint32_t *reading);
+void Amperes_Task(void *pvParameters);
```
add [doc comments](https://cloud.wikis.utexas.edu/wiki/spaces/LHRSOLAR/pages/407176278/Software+Design+Structure#%F0%9F%93%9C-Documentation) :)

Rav4s commented on Firmware/Drivers/Src/Amperes.c:
```diff
@@ -0,0 +1,141 @@
+StackType_t amperes_task_stack[configMINIMAL_STACK_SIZE];
+
+// Queue
+#define QUEUE_LENGTH 100
```
can u add reasoning for queue length

Lakshay983 replied:
QUEUE_LENGTH is not a great name since you may have other queues at some other point.

Also, make this define (and any others you see fit) overridable

```
ifndef QUEUE_LENGTH
#define QUEUE_LENGTH 100
#endif
```
allows you to at compile time (via a makefile) override the value for the sake of quick testing without changing the code.

Rav4s commented on Firmware/Drivers/Src/Amperes.c:
```diff
@@ -0,0 +1,141 @@
+static StaticQueue_t xStaticQueue;
+
+
+// static void ErrorHandler() {
+//     while (1) {}
+// }
+// static void successHandler(void) {
+//     // blinky
+//     while(1){
+//         HAL_GPIO_TogglePin(GPIOA, GPIO_PIN_5);
+//         HAL_Delay(500);
+//     }
+// }
```
i'd add these to like a `common.c` in Apps

Rav4s commented on Firmware/Drivers/Src/Amperes.c:
```diff
@@ -0,0 +1,141 @@
+    // Initialize queue
+    amperes_q = xQueueCreateStatic(QUEUE_LENGTH, ITEM_SIZE, qStorage, &xStaticQueue);
+
+    // Initialize CAN
```
will be different between the F4/L4 as well but you can use smth similar to [active-precharge](https://github.com/lhr-solar/Active-Precharge-Board/blob/firmware-fsgp-2025/Firmware/Drivers/Src/CANbus.c)

Rav4s commented on Firmware/Drivers/Src/Amperes.c:
```diff
@@ -0,0 +1,141 @@
+    if (adc_init(init, hadc1) != ADC_OK) return INIT_FAIL;
+    
+    // Set LED for debug
+    HAL_GPIO_WritePin(GPIOA, GPIO_PIN_5, GPIO_PIN_SET);
```
maybe have like a debug flag for all the LED stuff

Rav4s reviewed (changes requested):
Looks rlly good overall, just had a few questions/suggestions. as for packing the CAN message, you should be good to send MSB first (little endian) but @Lakshay983 would know best. we should document this on the BPS internal dbc soon

Lakshay983 commented on Firmware/Drivers/Inc/Amperes.h:
```diff
@@ -0,0 +1,34 @@
+    INIT_FAIL,
+    READING_OK,
+    READING_FAIL
+} AmperesStatus_t;
```
Feels like an odd set of statuses. Consider where all Amperes will fail / the states it will be in, and put all that into the enum

Lakshay983 commented on Firmware/Drivers/Src/Amperes.c:
```diff
@@ -0,0 +1,141 @@
+#include "Amperes.h"
+
+// Task
```
very blunt comments lmao

Lakshay983 commented on Firmware/Drivers/Src/Amperes.c:
```diff
@@ -0,0 +1,141 @@
+        .Pin = GPIO_PIN_1
+    };
+
+    HAL_GPIO_Init(GPIOA, &input);
```
it's good to keep these pin defintions into seperate defines

ie 
```
#define ADC_PORT GPIOA
#define ADC_PIN GPIO_PIN_1
```
so the pins feel less hard coded.

Lakshay983 commented on Firmware/Drivers/Src/Amperes.c:
```diff
@@ -0,0 +1,141 @@
+    init.OversamplingMode = DISABLE;
+    #endif
+
+    #ifdef STM32F4xx
```
again, imo not too useful to support both L4 and F4, we only need F4 for this

Embedded sharepoint code needs to support both but Amperes doesn't need too

Lakshay983 commented on Firmware/Drivers/Src/Amperes.c:
```diff
@@ -0,0 +1,141 @@
+    // ADC Handle
+    // amperes_adc.Instance = ADC1;
+    // ADC Init
+    init.ClockPrescaler = ADC_CLOCK_ASYNC_DIV1;
```
add comments for this stuff por favor

Lakshay983 commented on Firmware/Tests/test_can_loopback.c:
```diff
@@ -0,0 +1,6 @@
+#include "stm32xx_hal.h"
+
+int main() {
+
+    return 0;
+}
```
pls write tests

Lakshay983 commented on Firmware/Drivers/Src/Amperes.c:
```diff
@@ -0,0 +1,141 @@
+    while (1) {
+        // Get reading
+        uint32_t reading = 0;
+        Amperes_GetReading(&reading);
```
remember ur not just reading the ADC value and transmitting that onto CAN, you need to convert it to a current reading on ur side.

Lakshay983 commented on Firmware/Drivers/Src/Amperes.c:
```diff
@@ -0,0 +1,141 @@
+        // Debug
+        HAL_GPIO_TogglePin(GPIOA, GPIO_PIN_5);
+
+        vTaskDelay(200);
```
rn you're just polling the ADC, what you should be doing is blocking your task on the queue, waking up your task , processing the data, and then transmitting that data onto the CAN bus

There are a few benefits to this approach:
- much lower latency (when ADC data is ready to be processed we're there immediately instead of acting in only terms of your delay
- The RTOS can do other things 

Cons:
- you have no assurance that your task will ever start again (remember this is why we have a watchdog in BPS)
- ADC interrupts will occur at a somewhat consistent frequency but it'll be off and increase your jitter (the measure of the max and min time difference in doing something). More jitter means your system has more inconsistent timing which can mess with other things

Maybe you could have a timer that's happening in the background transmitting CAN messages at a consistent rate and then your Amperes task is blocking and waiting for data to appear in the ADC

As you write this code, consider all the weird wacky edge cases that could occur (given how long this board is on for they will all occur).   
So consider what happens if your adc messes up and loses data or stops responding. What if some other proces in your RTOS messes with this task.

https://freertos.org/Documentation/02-Kernel/02-Kernel-features/02-Queues-mutexes-and-semaphores/01-Queues#:~:text=Blocking%20on%20Queues,-Queue%20API%20functions&text=When%20a%20task%20attempts%20to,or%20the%20block%20time%20expires.

Lakshay983 reviewed (changes requested):
Looks good overall!

I'd move around the structure of ur code, like having the tasks inside Apps and then driver level code in drivers. I like having all your logic in tasks and low level embedded calls in drivers. See how the current BPS does it.

Make sure you test this incrementally as you go on hardware, it's a somewhat big waste of ur time to write a bunch of embeded code and not actually test it. Break things into testable chunks (this iis where the Tests directory comes into play).

micade145 committed:
finally back
- deleted hardware stuff
- better organization (Tasks folder, common.h)
- will address more comments later

micade145 committed:
updated Amperes Init

micade145 committed:
adc blinky

micade145 commented:
/gemini review

micade145 commented:
@gemini-code-assist review

micade145 commented:
gemini please

micade145 committed:
adc to current conversion, need to test

micade145 committed:
confirmed adc current reading hooray

micade145 committed:
started CAN stuff

micade145 committed:
i've added so much stuff so long ago and I don't know what works anymore

micade145 committed:
adc conversion on hardware timer6 interrupt (not tested)

micade145 committed:
new dataflow: hardware timer interrupt -> adc conversion -> adc task  -> queue -> CAN task

micade145 committed:
CAN message struct for storing adc value and current data

micade145 committed:
tweaked adc init for timer or no timer; need to get some fixed point stuff going for current conversion

micade145 committed:
comments

micade145 committed:
more ADC stuff, will probably not use hardware timer

micade145 committed:
added makefile flash-uart and debug thanks lg

micade145 committed:
fixed makefile i think

micade145 committed:
deleted timer interrupt for ADC; make sure ADC IRQ has prio >= 5

micade145 committed:
new Amperes_Task combines ADC_Task and CAN_Task; debugging w LEDs works, need to confirm values for ADC

micade145 committed:
eod 2/14/26; need to add printf to test ADC values; deleted CAN and ADC tasks (now consoldiated in Amperes_Task)

micade145 committed:
fixed point math for adc to current conversion

micade145 committed:
printf for debugging

Rav4s commented on .gitignore:
make sure gitignore covers both hw and fw

Rav4s commented on .gitmodules:
make sure this has kicad-sharepoint, merge with your hardware one

Rav4s commented on README.md:
add your conflu link + instructions on how to build/test

Rav4s commented on Firmware/Core/Inc/common.h:
```diff
@@ -0,0 +1,38 @@
+#ifndef COMMON_H
+#define COMMON_H
+/**
```
standardize to #pragma once, instead of the #ifndef include guards to keep consistent across our codebase

Rav4s commented on Firmware/Core/Inc/common.h:
```diff
@@ -0,0 +1,38 @@
+ */
+void error_handler();
+
+void Error_Handler();
```
why both of these? should just use the caps one (`Error_Handler()` and `Success_Handler()`) since they will override stm32 weakly defined fns

Rav4s commented on Firmware/Core/Src/common.c:
```diff
@@ -0,0 +1,139 @@
+    // Turn on fault LED
+    HAL_GPIO_WritePin(AMPERES_GPIO_PORT, AMPERES_FAULT_PIN, GPIO_PIN_SET);
+    // Disable interrupts ???
+    __disable_irq();
```
```suggestion
```

We should keep interrupts enabled in error state so the scheduler keeps running, and we can keep sending/receiving CAN messages

Rav4s commented on Firmware/Drivers/Inc/Amperes.h:
```diff
@@ -0,0 +1,182 @@
+ *  - Gain: 100 * 0.4 * 2.49 V/V = 99.6 V/V
+ *  - Reference voltage: 1.25 V (for bidirectinal sense amp)
+ *  - ADC: 12 bit
+ * ================================================================ */
```
this is awesome :)

Rav4s commented on Firmware/Drivers/Inc/Amperes.h:
might be more logical to break up into Amperes_ADC and Amperes_CAN files. thoughts?

Rav4s commented on Firmware/Drivers/Inc/Amperes.h:
```diff
@@ -0,0 +1,182 @@
+ *  - Current range: -46 to +78 A (0.1 - 3.2 V respectively)
+ *  - Shunt: 250 uOhm
+ *  - Gain: 100 * 0.4 * 2.49 V/V = 99.6 V/V
+ *  - Reference voltage: 1.25 V (for bidirectinal sense amp)
```
```suggestion
 *  - Reference voltage: 1.25 V (for bidirectional sense amp)
```

Rav4s commented on Firmware/Drivers/Inc/Amperes.h:
```diff
@@ -0,0 +1,182 @@
+#define AMPERES_SAMPLE_TIME ADC_SAMPLETIME_47CYCLES_5
+// ADC_SAMPLETIME_2CYCLES_5
+#define AMPERES_ADC_PORT    GPIOA
+#define AMPERES_ADC_PIN     GPIO_PIN_0
```
i would put these in a pinDefs.h with all your pin defines

Rav4s commented on Firmware/Drivers/Inc/Amperes.h:
```diff
@@ -0,0 +1,182 @@
+/** ================================================================
+ *  CAN
+ * ================================================================ */
+#define AMPERES_STD_ID 0x1
```
eventually we'll wanna grab this from the DBC, but thats not merged yet so it's chill for now

Rav4s commented on Firmware/Drivers/Inc/Amperes.h:
```diff
@@ -0,0 +1,182 @@
+ *  Amperes Functions
+ * ================================================================ */
+
+typedef enum AmperesStatus {    // TODO: better states
```
address TODO lol

Rav4s replied:
if you want, statuses seem okay to me

Rav4s commented on Firmware/Drivers/Src/Amperes.c:
```diff
@@ -0,0 +1,219 @@
+
+}
+
+static bool Amperes_ADC_Init() {
```
since this is static, i'd add a short comment here explaining the function + when it could fail

Rav4s commented on Firmware/Drivers/Src/Amperes.c:
```diff
@@ -0,0 +1,219 @@
+    return true;
+}
+
+static bool Amperes_CAN_Init() {
```
same for this one

Rav4s commented on Firmware/Drivers/Src/Amperes.c:
```diff
@@ -0,0 +1,219 @@
+    tx_header.StdId = AMPERES_STD_ID;
+    tx_header.RTR = CAN_RTR_DATA;
+    tx_header.IDE = CAN_ID_STD;
+    tx_header.DLC = 8;
```
```suggestion
    tx_header.DLC = 6;
```
If you're only using 6 bytes, the DLC can be 6

Rav4s commented on Firmware/Drivers/Src/Amperes.c:
```diff
@@ -0,0 +1,219 @@
+    // Split data into uint8 elements; LSB at index 0
+
+    /* Raw ADC Value: uint16_t */
+    uint8_t tx_data[8] = {0};
```
```suggestion
    uint8_t tx_data[6] = {0};
```

Rav4s commented on Firmware/Drivers/Src/Amperes.c:
```diff
@@ -0,0 +1,219 @@
+    tx_data[2] = (uint8_t) (data->current_data & 0xFF);
+    tx_data[3] = (uint8_t) ((data->current_data >> 8) & 0xFF);
+    tx_data[4] = (uint8_t) ((data->current_data >> 16) & 0xFF);
+    tx_data[5] = (uint8_t) ((data->current_data >> 24) & 0xFF);
```
eventually we'll switch this over to dbc generated shifting/masking as well, just a note to remember

Rav4s commented on Firmware/Core/Src/main.c:
```diff
@@ -0,0 +1,30 @@
+    // Start scheduler
+    vTaskStartScheduler();
+
+    error_handler();
```
```suggestion
    Error_Handler();
```

Rav4s commented on Firmware/Tasks/Inc/Tasks.h:
```diff
@@ -0,0 +1,64 @@
+#define TASKS_H
+#include "Amperes.h"
+
+// TODO: SET STACK SIZE AND PRIORITIES
```
address TODO

Rav4s commented on Firmware/Tasks/Inc/Tasks.h:
```diff
@@ -0,0 +1,64 @@
+/**
+ * Init Task
+ * - Lowest priority: initializes tasks and then deletes itself
+ * - TODO: I'm only initializing one task do I even need this mane
```
not rlly but in case you add more its nice to have lmao

Rav4s commented on Firmware/Tasks/Inc/Tasks.h:
```diff
@@ -0,0 +1,64 @@
+void Amperes_Task(void *pvParameters);
+
+/**
+ * @brief Initializes Amperes and all tasks (ADC and CAN),
```
```suggestion
 * @brief Initializes Amperes hardware and task,
```

Rav4s commented on Firmware/Tasks/Src/Amperes_Task.c:
```diff
@@ -0,0 +1,74 @@
+    uint8_t counter = 0;
+    uint8_t conv_count = 0;
+
+    Init_UART_Printf();
```
shouldn't this be in Init_Task?

Rav4s commented on Firmware/Tasks/Src/Amperes_Task.c:
```diff
@@ -0,0 +1,74 @@
+    Init_UART_Printf();
+
+    // Temporarily using CAN RX pin for 
+    // profiling loop time w logic analyzer
```
LOL

Rav4s commented on Firmware/Tasks/Src/Amperes_Task.c:
```diff
@@ -0,0 +1,74 @@
+    __HAL_RCC_GPIOB_CLK_ENABLE();
+    HAL_GPIO_Init(GPIOB, &debug_gpio);
+
+    // Loop seems to take around 24 us
```
🤏

Rav4s commented on Firmware/Tasks/Src/Amperes_Task.c:
```diff
@@ -0,0 +1,74 @@
+    // Loop seems to take around 24 us
+    while (1) {
+        HAL_GPIO_TogglePin(GPIOB, GPIO_PIN_8);  // toggle CAN RX for measuring loop time
+        HAL_GPIO_TogglePin(GPIOB, GPIO_PIN_8);  // toggle CAN RX for measuring loop time
```
valvano would approve

Rav4s commented on Firmware/Tasks/Src/Amperes_Task.c:
```diff
@@ -0,0 +1,74 @@
+        // Start ADC reading. Clear queue because xQueueReceive with
+        // portMAX_DELAY blocks until queue goes from empty -> has data.
+        if (Amperes_StartADC(true) != AMPERES_OK) { 
+            // error_handler();  // TODO: handle errors
```
address TODO

Rav4s commented on Firmware/Tasks/Src/Amperes_Task.c:
```diff
@@ -0,0 +1,74 @@
+        if (Amperes_GetReading(&message, portMAX_DELAY) == AMPERES_OK) {
+            conv_count++;
+            adc_sum += message.adc_voltage;         // 10 sums of unsigned 12b values can fit into uint16_t
+            current_sum += message.current_data;    // can probably fit yea
```
easily 😅

Rav4s commented on Firmware/Tasks/Src/Amperes_Task.c:
```diff
@@ -0,0 +1,74 @@
+            conv_count++;
+            adc_sum += message.adc_voltage;         // 10 sums of unsigned 12b values can fit into uint16_t
+            current_sum += message.current_data;    // can probably fit yea
+            HAL_GPIO_TogglePin(AMPERES_GPIO_PORT, AMPERES_CHARGE_PIN);
```
do you have somehwere for setting charge/discharge leds based on the actual value?

Rav4s commented on Firmware/Tasks/Src/Amperes_Task.c:
```diff
@@ -0,0 +1,74 @@
+        // xLastWakeTime is updated within vTaskDelayUntil.
+        counter++;
+        HAL_GPIO_TogglePin(GPIOB, GPIO_PIN_8);  // toggle CAN RX for measuring loop time
+        vTaskDelayUntil(&xLastWakeTime, AMPERES_TASK_PERIOD); 
```
nice!

Rav4s commented on Firmware/Tests/test_adc.c:
```diff
@@ -0,0 +1,88 @@
+//         xQueueSend(adc_queue, &val, 0);
+//         vTaskDelay(pdMS_TO_TICKS(1000));
+//     }
+// }
```
if you're not using the queue stuff just delete

Rav4s commented on Firmware/Tests/test_adc_blinky.c:
```diff
@@ -0,0 +1,79 @@
+    GPIO_InitTypeDef debug_gpio = {
+        .Mode = GPIO_MODE_OUTPUT_PP,
+        .Pull = GPIO_NOPULL,
+        .Pin = GPIO_PIN_8 | GPIO_PIN_9
```
use your pindefs from header file

Rav4s reviewed (changes requested):
Nice job! Overall very polished and well written :)

Make sure to merge in main so you have the Rev B hardware changes on this branch as well, so it doesn't remove those when you merge this PR.

micade145 committed:
merge main into firmware for hardware changes

micade145 committed:
updated readme, error handler stuff

micade145 committed:
Merge remote-tracking branch 'origin/main' into Firmware

micade145 committed:
fixed adc - > current conversion

micade145 committed:
addressed some comments

micade145 committed:
eod 2/21: led stuff, working on CAN

micade145 committed:
added sConfig for ADC, CAN msp init

micade145 committed:
might be back (can)

micade145 committed:
file for msp inits, all tests build

micade145 commented:
/gemini review

micade145 committed:
smth

micade145 committed:
cooler conversion (works for +-10A through the shunt, within ~50mA)

micade145 committed:
7

micade145 commented:
/gemini review

micade145 committed:
addressed comment and concerns

micade145 committed:
documentation, make sure CAN mode is normal?

Lakshay983 commented on Firmware/Drivers/Inc/Amperes.h:
```diff
@@ -0,0 +1,153 @@
+ *  - ADC: 12 bit
+ * ================================================================ */
+
+/* LED Pins */
```
would like this stuff moved to a different header file.
Like a pindef file

Lakshay983 commented on Firmware/Drivers/Inc/Amperes.h:
```diff
@@ -0,0 +1,153 @@
+#define AMPERES_MSG_DLC     6
+#define AMPERES_CAN_PORT    GPIOB
+#define AMPERES_RX_PIN      GPIO_PIN_8
+#define AMPERES_TX_PIN      GPIO_PIN_9
```
same here for moving to other file.

Also, presently those pins are hard coded in your MSPInit, you should use these defines in the msp_inits

Lakshay983 commented on Firmware/Drivers/Src/Amperes.c:
```diff
@@ -0,0 +1,210 @@
+    if (can_init(hcan1, &sFilterConfig) != CAN_OK) return false;
+    
+    return true;
+}
```
can you move the CAN stuff to a diff file

Lakshay983 commented on Firmware/Drivers/Src/Amperes.c:
```diff
@@ -0,0 +1,210 @@
+    if (!Amperes_CAN_Init()) return AMPERES_INIT_FAIL;
+
+    // Start CAN
+    if (can_start(hcan1) != CAN_OK) return AMPERES_INIT_FAIL;
```
these should be called in the init task

Lakshay983 commented on Firmware/Tasks/Src/Amperes_Task.c:
```diff
@@ -0,0 +1,79 @@
+                counter = conv_count = 0;
+            } else {
+                // Handle error: no adc conversions
+                // reinit ADC?
```
maybe add like a printf for adc_errors

Lakshay983 commented on Firmware/Tasks/Src/Amperes_Task.c:
```diff
@@ -0,0 +1,79 @@
+                // reinit ADC?
+            }
+            
+            Amperes_UpdateLEDs(message.current_data);
```
from a high level readability standpoint, i'd either move the stauts led code directly into this function of explain in comments what it's doing.
i'd say move the code directly into the while loop so i can look at this thread and see exactly what the intended behavior of the lights are

Lakshay983 commented on Firmware/Tasks/Src/TaskInit.c:
```diff
@@ -0,0 +1,29 @@
+StackType_t     Task_Init_Stack[TASK_INIT_STACK_SIZE];
+
+void Task_Init() {
+    // Initialize Amperes GPIO, ADC, and CAN
```
i would recommend having the status leds in a different file

Lakshay983 commented on Firmware/Makefile:
```diff
@@ -0,0 +1,128 @@
+	$(MAKE) -C $(BUILD_MAKEFILE_DIR) debug
+	
+# Help
+.PHONY: help
```
look at one of my recent commits to the embedded-sharepoint makefile, i added -e to all the echos so colors work more consistently

Lakshay983 reviewed (changes requested):
mostly semantics changes, but very close to being done!

I talked about this in the slack thread, but a few feature requests:

-  mirror can messages over USB
- implement the idle task
^ this stuff should be for different PRs tho

micade145 committed:
mane clock config what

micade145 committed:
config header file

micade145 committed:
makefile change

micade145 committed:
before I refactor stuff

micade145 committed:
refactored Amperes.c into CAN and ADC files

micade145 committed:
sum light

micade145 committed:
normal can mode so i don't forget

Rav4s reviewed (approved):
LGTM
